### ISSUE TITLE
[Question] Intercept java method invocations

### ISSUE THREAD

**Issue by Brunomachadob** [beanshell/beanshell#35]
Hi everyone, I'm studying beanshell to make a simple console for my application server, so me or a certified user could run scripts, get or set data and another things at runtime.
I'm loving the library and suited all my needs.
For now I have only one requirement left, I need to include a authentication mechanism to my console.
The simple way out is simply do the auth before open the console, but this would  prohibit my user to do simple tasks like get some sort of data, and depending on the data, this doesn't need auth.
What I want to do is intercept java method invocations in classes that are deployed in my server (similar if I was using a Proxy on them and overriding the invoke method), so I could trigger or not a auth prompt for that call in particular.
There is a way to achieve this or at least something similar?

Here is a hypothetic example: 
I have a singleton with private static variables (caches for example) and some public methods.
If my application call those methods, it doesn't need an authentication, but called from my console, it needs. 

If the functionality does not exist, I am willing to study the possibility of implementing and sending a pull request!

Sorry for my bad english haha.
Thanks in advance.

**Comment by Brunomachadob** [comment]
After playing around a little with the bsh source, i achieve this (for test purposes only, I now we can do better):

```java
package bsh.interceptors;
public interface MethodInvokeInterceptor {
   public void intercept(Object instance, Method method, Object[] args) throws RuntimeException;
}
```
```java
//Main.class
interpreter.setMethodInvokeInterceptor(new MethodInvokeInterceptor() {
   private Set<String> authMethods = new HashSet<String>() {
     {
	add("java.lang.System.setProperty");
     }
  };
       @Override
	public void intercept(Object instance, Method method, Object[] args) throws RuntimeException {

	if (authMethods.contains(method.getDeclaringClass().getName() + "." + method.getName())) {
		throw new AuthException("Auth needed");
	}

  }
});
``` 
And then, I catch the AuthException and launch the auth pop up.

**Comment by nickl-** [comment]
This is one way to address the issue another way might be to limit the methods you have access to by limiting the classpath and the reach of discovering new methods into the namespace.

Before we can invoke a method we first have to locate it, whether that means creating the class with bytecode or loading an existing class and reflecting the methods. If you don't have access to any new methods then you can't invoke them.

We could probably do something like secured namespaces, preloaded with the methods and variables granted and then requiring some sort of authentication/authorisation to gain access to the namespace.

**Comment by stefanofornari** [comment]
Maybe another method could be to install your own security manager and
protect your methods by it. Your security manager could trigger the
authentication request.

HTH

Ste

**Comment by nickl-** [comment]
That sounds promising! Great idea @stefanofornari !!

I haven't done this before... do you have an example handy or is it a big mission?

Perhaps we should add a reference implementation I could see this be a useful addition.

**Comment by ecki** [comment]
BTW just want to mention, since security manager is now deprecated the secured namespace approach sounds more reasonable. Having said that the method interceptor seems like a useful API extension in all cases. Also for tracing or even method extension,

**Comment by jimjag** [comment]
Closing during a backlog triage pass: this reads as a question already answered with workarounds in the thread, and was labeled out of scope for v3.0 with no concrete ask remaining. Reopen if this is still relevant.
